rllib/utils/nested_dict.py

```diff
@@ -0,0 +1,305 @@
+
+        Args:
+            k: the key to get. This can be a string or a sequence of strings.
+            default: the default value to return if `k` is not in the `NestedDict`. If
+                default is `None`, and `k` is not in the `NestedDict`, a `KeyError` is
+                raised.
+
+        Returns:
+            The value of `self[k]`.
+
+        Raises:
+            KeyError: if `k` is not in the `NestedDict` and default is None.
+        """
+        k = _flatten_index(k)
+
+        if k not in self:
+            if default is not None:
+                return default
+            else:
+                raise KeyError(k)
+
+        k = _flatten_index(k)
+        data_ptr = self._data
+        for key in k:
+            # This is to avoid the recursion on __getitem__
+            if isinstance(data_ptr, NestedDict):
+                data_ptr = data_ptr._data
+            data_ptr = data_ptr[key]
+        return data_ptr
+
+    def __getitem__(self, k: SeqStrType) -> T:
+        output = self.get(k)
+        if isinstance(output, NestedDict):
+            raise IndexError("Use get for partial indexing.")
+        return output
+
+    def __setitem__(self, k: SeqStrType, v: Union[T, _NestedMappingType]) -> None:
+        """This is a zero-copy operation. The pointer to value if preserved in the
+        internal data structure."""
+        if isinstance(v, Mapping) and len(v) == 0:
+            return
+        if not k:
+            raise IndexError("Use valid index value.")
+        k = _flatten_index(k)
+        v = NestedDict(v) if isinstance(v, Mapping) else v
+        data_ptr = self._data
+        for k_indx, key in enumerate(k):
+            # this is done to avoid recursion over __setitem__
+            if isinstance(data_ptr, NestedDict):
+                data_ptr = data_ptr._data
+            if k_indx == len(k) - 1:
+                data_ptr[key] = v
+            elif key not in data_ptr:
+                data_ptr[key] = NestedDict()
+            data_ptr = data_ptr[key]
+
+    def __iter__(self) -> Iterator[SeqStrType]:
+        data_ptr = self._data
+        # do a DFS to get all the keys
+        stack = [((StrKey(k),), v) for k, v in data_ptr.items()]
+        while stack:
+            k, v = stack.pop(0)
+            if isinstance(v, NestedDict):
+                stack = [(k + (StrKey(k2),), v) for k2, v in v._data.items()] + stack
+            else:
+                yield tuple(k)
+
+    def __delitem__(self, k: SeqStrType) -> None:
+        if k not in self:
+            raise KeyError(k)
+
+        k = _flatten_index(k)
+        # for deletion we need backtracking. Recursion is much simpler to implement.
+        # we don't expect that much of deletion, so we can afford to do recursion.
+        if len(k) == 1:
+            del self._data[k[0]]
+        else:
+            del self._data[k[0]][k[1:]]
+            if not self._data[k[0]]:
+                del self._data[k[0]]
```

Comment:
I feel like backtracking this isn't that much more complicated, something like:
```
ks = []
ns = []
root = self._data
for k in _flatten_index(k):
    if k not in root._data:
        raise KeyError(str(ks + [k]))
    ks.append(k)
    ns.append(root)
    root = root._data

del ns[-1][ks[-1]]

for i in reverse(range(len(ks) - 1)):
    if not ns[i + 1]:
        del ns[i][ks[i]]
```
anyways, up to you.

Reply:
fixed.